desktopexporter/app/components/waterfall-view/waterfall-view.tsx

```diff
@@ -24,13 +25,53 @@ export function WaterfallView(props: WaterfallViewProps) {
   const spanNameColumnWidth = 300;
   const serviceNameColumnWidth = 200;
 
+  let { orderedSpans, traceTimeAttributes, selectedSpanID, setSelectedSpanID } =
+    props;
+
+  // Set up keyboard navigation
+  let arrowUpPressed = useKeyPress("ArrowUp");
+  let arrowDownPressed = useKeyPress("ArrowDown");
+  let kPressed = useKeyPress("k");
+  let jPressed = useKeyPress("j");
+
+  let selectedIndex = orderedSpans.findIndex(
+    (span) => span.metadata.spanID === selectedSpanID,
+  );
+  let firstSelectableIndex = orderedSpans.findIndex(
+    (span) => span.status === SpanDataStatus.present,
+  );
+
+  useEffect(() => {
+    if (arrowUpPressed || kPressed) {
+      // Move up while skipping the missing spans in incomplete traces
+      if (selectedIndex > firstSelectableIndex) {
+        do {
+          selectedIndex--;
+        } while (orderedSpans[selectedIndex].status === SpanDataStatus.missing);
+        setSelectedSpanID(orderedSpans[selectedIndex].metadata.spanID);
+      }
+    }
+  }, [arrowUpPressed, kPressed]);
+
+  useEffect(() => {
+    if (arrowDownPressed || jPressed) {
+      // Move down while skipping the missing spans in incomplete traces
+      if (selectedIndex < orderedSpans.length - 1) {
+        do {
+          selectedIndex++;
+        } while (orderedSpans[selectedIndex].status === SpanDataStatus.missing);
+        setSelectedSpanID(orderedSpans[selectedIndex].metadata.spanID);
+      }
+    }
+  }, [arrowDownPressed, jPressed]);
+
   let rowData = {
-    orderedSpans: props.orderedSpans,
-    traceTimeAttributes: props.traceTimeAttributes,
+    orderedSpans: orderedSpans,
+    traceTimeAttributes: traceTimeAttributes,
     spanNameColumnWidth: spanNameColumnWidth,
     serviceNameColumnWidth: serviceNameColumnWidth,
-    selectedSpanID: props.selectedSpanID,
-    setSelectedSpanID: props.setSelectedSpanID,
+    selectedSpanID: selectedSpanID,
+    setSelectedSpanID: setSelectedSpanID,
```

Comment:
JS has key value "punning" which means there is a shorter way to write this when the keys match the variable name.

```typescript
let rowData = {
    orderedSpans,
    traceTimeAttributes,
    spanNameColumnWidth,
    serviceNameColumnWidth,
    selectedSpanID,
    setSelectedSpanID,
  };
```